lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+/*
+  This file is part of Leela Chess Zero.
+  Copyright (C) 2018 The LCZero Authors
+
+  Leela Chess is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  Leela Chess is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "network_cudnn.h"
+#include "utils/bititer.h"
+#include <functional>
+#include <cassert>
+#include <mutex>
+
+#include <cublas_v2.h>
+#include <cudnn.h>
+
+
+void cudnnError(cudnnStatus_t status, const char* file, const int& line)
+{
+    if (status != CUDNN_STATUS_SUCCESS)
+    {
+        printf("CUDNN error: %s (%s:%d)\n", cudnnGetErrorString(status), file, line);
+        getchar();
+        exit(1);
+    }
+}
+
+#define reportCUDNNErrors(status)  cudnnError(status, __FILE__, __LINE__)
+
+
+// 256 MB fixed scratch memory size (hardcoded for now)
+#define CUDA_SCRATCH_SIZE (256*1024*1024)
```

Comment:
It's better to have C++ constants rather than macros for macros. (unless CUDA requires it so)
Also it's encouraged to put them into anonymous namespace rather than make static.

Reply:
This should probably be calculated to avoid wasting GPU memory. See cudnnGetConvolutionForwardWorkspaceSize function.